pkg/pool-weighted/contracts/smart/ManagedPool.sol

```diff
@@ -444,6 +445,268 @@ contract ManagedPool is BaseWeightedPool, ProtocolFeeCache, ReentrancyGuard {
+    function _joinAddToken(
+        IERC20[] memory tokens,
+        uint256 tokenIndex,
+        uint256 tokenAmountIn,
+        uint256[] memory maxAmountsIn,
+        address recipient
+    ) private {
+        getVault().joinPool(
+            getPoolId(),
+            address(this),
+            payable(recipient),
+            IVault.JoinPoolRequest({
+                assets: _asIAsset(tokens),
+                maxAmountsIn: maxAmountsIn,
+                userData: abi.encode(WeightedPoolUserData.JoinKind.ADD_TOKEN, tokenIndex, tokenAmountIn),
+                fromInternalBalance: false
+            })
+        );
+    }
+
+    /**
+     * @dev 1) Validate the operation (and calculate the new weightSum and bptAmountOut to return)
+     *         - the incoming normalizedWeight is valid
+     *         - adding a token will not exceed the token limit
+     *         - there is no ongoing or pending weight change
+     *         - adding the new token at the given weight does not lower any other weights below the minimum
+     *         - the final BPT price is at or above the calculated minimum
+     *      2) Register the new token, with the given asset manager (and return the final sorted token list,
+     *         and index of the new token). Note that the token order can completely change. Note that after
+     *         registration and before joining, the pool is in an invalid state, with a zero invariant.
+     *         - register the token with the Vault
+     *         - adjust the management fees data structure to the new token order (preserving any uncollected fees)
+     *         - adjust the rest of the token state (including token count)
+     *      3) Join the pool, transferring tokens to the Vault, and restoring the pool to functional status
+     *      4) Finally, update the stored weightSum, and return the bptAmountOut. The caller may then mint BPT,
+     *         depending on the use case.
+     */
+
+    function _addToken(
+        IERC20 token,
+        uint256 normalizedWeight,
+        uint256 tokenAmountIn,
+        address assetManager,
+        uint256 minBptPrice,
+        address sender,
+        address recipient
+    ) internal returns (uint256) {
+        (uint256 weightSumAfterAdd, uint256 bptAmountOut) = _validateAddToken(
+            token,
+            normalizedWeight,
+            tokenAmountIn,
+            minBptPrice
+        );
+
+        (IERC20[] memory tokens, uint256 tokenIndex, uint256[] memory maxAmountsIn) = _registerNewToken(
+            token,
+            normalizedWeight.mulUp(weightSumAfterAdd),
+            tokenAmountIn,
+            assetManager
+        );
+
+        // Transfer tokens from the sender to this contract, since the sender for the join must be the pool
+        token.transferFrom(sender, address(this), tokenAmountIn);
+        token.approve(address(getVault()), tokenAmountIn);
+
+        _joinAddToken(tokens, tokenIndex, tokenAmountIn, maxAmountsIn, recipient);
```

Comment:
Let's inline this to be consistent with `removeToken`

Reply:
This causes stack too deep issues but this might be fixed by some later changes.